content/articles/2019-02-purrr-0-3-0.Rmd

```diff
@@ -0,0 +1,323 @@
+reduce(1:100, ~ if (.x > 50) done(.x) else .x + .y)
+```
+
+This feature takes inspiration from the [Clojure](https://clojuredocs.org/clojure.core/reduced) language.
+
+
+## Consistency
+
+In this polishing release, a lot of effort went towards consistency of behaviour and of the interface.
+
+
+### Behaviour
+
+#### `modify()` now wraps `[[<-`
+
+An important improvement is that `modify()` is now a wrapper around `[[<-` rather than `[<-`. This should make it directly compatible with a larger set of vector classes.
+
+Thanks to the work of Mikko Marttila (@mikmart), `pmap()` and `pwalk()` do a better job of preserving S3 classes. `pluck()` also works better with S3 objects. In the next version of purrr, we plan to use the in-development vctrs package to provide more principled and predictable vector operations.
+
+
+#### Stricter predicate checking
+
+purrr now checks the results of your predicate functions, which must now consistently return `TRUE` or `FALSE`. We no longer offer support for `NA` or for boolish numeric values (R normally interprets 0 as `FALSE` and all other values as `TRUE`). The purpose of this change is to detect errors earlier with a more relevant error messages.
+
+```{r, error = TRUE}
+keep(c(1, NA, 3), ~ . %% 2 == 0)
+```
+
+
+### Interface
+
+#### Direction of application
+
+The direction of application is now specified the same way across purrr functions. `reduce()`, `compose()` and `detect()` now have a `.dir` parameter that can take the value `"forward"` or `"backward"`. This terminology should be less ambiguous than "left" and "right":
+
+```{r, eval = FALSE}
+reduce(1:4, `-`, .dir = "backward")
+
+compose(foo, bar, .dir = "forward")
+
+detect(1:5, ~ . %% 2 == 0, .dir = "backward")
+```
+
+Note that the backward version of `reduce()` (called right-reduce in the literature) applies the reduced function in a slightly different way than `reduce_right()`. The new algorithm is more consistent with how this operation is usually defined in other languages.
+
+Following the introduction of the `.dir` parameters, the `_right` variants such as `reduce_right()` have been soft-deprecated, as well as the `.right` parameter of `detect()` and `detect_index()`.
+
+
+#### partial()
+
+`partial()` has been rewritten to be a simple wrapper around `call_modify()` and `eval_tidy()` from the rlang package. Consequently, the `.env`, `.lazy` and `.first` arguments are soft-deprecated and replaced by a flexible syntax.
+
+To control the timing of evaluation, unquote the partialised arguments that should be evaluated only once when the function is created. The non-unquoted arguments are evaluated at each invokation of the function:
+
+```{r}
+my_list <- partial(list, lazy = rnorm(3), eager = !!rnorm(3))
+
+my_list()
+
+my_list()
+```
+
+You can also control the position of the future arguments by passing an empty `... = ` parameter. This syntax is powered by `rlang::call_modify()` to control where the position of dots in a function call:
```

Comment:
Something's off with this sentence / description of `rlang::call_modify()`.

Reply:
oh yes, also it's not easy to explain that syntax succintly. Does this do a better job?


> You can also control the position of the future arguments by passing an empty `... = ` parameter. This syntax is powered by `rlang::call_modify()` and allows you to add or move dots in a quoted function call. In the case of `partial()`, the dots represent the future arguments. We use this syntax in the following snippet to position the future arguments right between two partialised arguments: